src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `src_namescol` : the column that will become the new header.
+  Defaults to first column.
+- `dest_namescol` : the name of the first column in the returned `DataFrame`.
+  Defaults to the same name as `src_namescol`.
+- `makeunique` : if `false` (the default), an error will be raised
+  if duplicate names are found; if `true`, duplicate names will be suffixed
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+Note: The eltypes of columns in resulting `DataFrame` will depend
+on the eltypes of _all_ input columns.
+That is, if the source `DataFrame` contains `String` and `Int` columns,
+all resulting columns will have eltype `Any`.
+If the source has a mix of numeric types (eg. `Float64` and `Int`),
+all columns in resulting `DataFrame` will be promoted to `Float64`.
+
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2));
+
+julia> df2 = DataFrame(a=["x", "y"], b=[1, "str"], c=[1,2], d=rand(Bool,2));
+
+julia> permutedims(df1) # note the column type
+3×3 DataFrame
+│ Row │ a      │ x        │ y        │
+│     │ String │ Float64  │ Float64  │
+├─────┼────────┼──────────┼──────────┤
+│ 1   │ b      │ 0.982197 │ 0.263357 │
+│ 2   │ c      │ 1.0      │ 2.0      │
+│ 3   │ d      │ 0.0      │ 1.0      │
+
+julia> permutedims(df2)
+3×3 DataFrame
+│ Row │ a      │ x   │ y   │
+│     │ String │ Any │ Any │
+├─────┼────────┼─────┼─────┤
+│ 1   │ b      │ 1   │ str │
+│ 2   │ c      │ 1   │ 2   │
+│ 3   │ d      │ 0   │ 0   │
+````
+"""
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                          dest_namescol::Union{Symbol, AbstractString};
+                          makeunique::Bool=false)
+
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
+    eltype(df[!, src_namescol]) <: SymbolOrString || throw(
+            ArgumentError("src_namescol must have eltype `Symbol` or `AbstractString`"))
+
+    df_notsrc = df[!, Not(src_namescol)]
+    df_permuted = DataFrame([names(df_notsrc)], [dest_namescol])
+
+    m = permutedims(Matrix(df_notsrc))
+    df_tmp = rename!(DataFrame(Tables.table(m)), df[!, src_namescol], makeunique=makeunique)
+    hcat!(df_permuted, df_tmp, copycols=false)
+end
+
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex;
+                          makeunique::Bool=false)
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(ArgumentError("`src_namescol` doesn't exist"))
+        dest_namescol = _names(df)[src_namescol]
+    else
+        dest_namescol = src_namescol
+    end
+    permutedims(df, src_namescol, dest_namescol; makeunique=makeunique)
+end
+
+function Base.permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+    permutedims(df, 1; makeunique=makeunique)
+end
```

Comment:
we typically use short form in such cases:
```suggestion
Base.permutedims(df::AbstractDataFrame; makeunique::Bool=false) =
    permutedims(df, 1; makeunique=makeunique)
```